### PR TITLE
AJ-955 - Adding Leonardo Client to WDS to extract source WDS Url during cloning

### PR DESCRIPTION
**Pull request by yuliadub**
This is a split out PR in hopes to make https://github.com/DataBiosphere/terra-workspace-data-service/pull/280 smaller. This PR adds the leonardo client that will be used to request the source WDS url to enable cloning. 

Reminder:

PRs merged into main will not automatically generate a PR in https://github.com/broadinstitute/terra-helmfile to update the WDS image deployed to kubernetes - this action will need to be triggered manually by running the following github action: https://github.com/DataBiosphere/terra-workspace-data-service/actions/workflows/tag.yml. Dont forget to provide a Jira Id when triggering the manual action, if no Jira ID is provided the action will not fully succeed. 

After you manually trigger the github action (and it completes with no errors), you must go to [the terra-helmfile](https://github.com/broadinstitute/terra-helmfile) repo and verify that this generated a PR that merged successfully.

The terra-helmfile PR merge will then generate a PR in [leonardo](https://github.com/DataBiosphere/leonardo).  This will automerge if all tests pass, but if jenkins tests fail it will not; be sure to watch it to ensure it merges. To trigger jenkins retest simply comment on PR with "jenkins retest".

### CODE REVIEW

**Review by ashanhol — APPROVED** [review]

File: service/src/main/java/org/databiosphere/workspacedataservice/leonardo/HttpLeonardoClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+public class HttpLeonardoClientFactory implements LeonardoClientFactory {
+
+    final String leoEndpointUrl;
+    private final OkHttpClient commonHttpClient;
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpLeonardoClientFactory.class);
+
+
+    public HttpLeonardoClientFactory(String leoUrl) {
+        this.leoEndpointUrl = leoUrl;
+        this.commonHttpClient = new org.broadinstitute.dsde.workbench.client.sam.ApiClient()
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/leonardo/HttpLeonardoClientFactory.java:25]: ooh, this should be
```java
this.commonHttpClient = new org.broadinstitute.dsde.workbench.client.leonardo.ApiClient()
```
or better, just
```java
this.commonHttpClient = new ApiClient()
```
it's tricky that all these generated clients use the same class names in different packages!

**Review by calypsomatic — APPROVED** [review]

File: service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    @Test
+    void testWdsUrlReturned() throws ApiException {
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("wds", AppStatus.RUNNING,1));
+        assertEquals(expectedUrl, url);
+    }
+
+    @Test
+    void testWdsUrlNotFoundWrongStatus() throws ApiException {
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("wds", AppStatus.DELETED, 1));
+        assertEquals(null, url);
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java:64]: from IntelliJ: 'assertEquals()' can be simplified to 'assertNull()'

File: service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Test
+    void testWdsUrlNotFoundWrongKey() throws ApiException {
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("not-wds", AppStatus.RUNNING, 1));
+        assertEquals(null, url);
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java:70]: from IntelliJ: 'assertEquals()' can be simplified to 'assertNull()'

File: service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    void testWdsUrlNotReturned() throws ApiException {
+        final int statusCode = HttpStatus.INTERNAL_SERVER_ERROR.value();
+        given(mockAppsApi.listAppsV2(String.valueOf(UUID.randomUUID()), "things", null, null))
+                .willThrow(new org.broadinstitute.dsde.workbench.client.leonardo.ApiException(statusCode, "Intentional error thrown for unit test"));
+        var exception = assertThrows(LeonardoServiceException.class, () -> leonardoDao.getWdsEndpointUrl(any()));
+        assertEquals(statusCode, exception.getRawStatusCode());
+    }
+
+    @Test
+    void testWdsUrlReturned() throws ApiException {
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java:56]: from IntelliJ: Exception 'org.broadinstitute.dsde.workbench.client.leonardo.ApiException' is never thrown in the method (this applies to all tests _except_ testWdsUrlNotReturned)

File: service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    @Test
+    void testWdsUrlMultiple() throws ApiException {
+        // tests the case if there are 2 running wds apps
+        var url = buildAppResponseAndCallExtraction(generateListAppResponse("wds", AppStatus.RUNNING,2));
+        assertEquals(expectedUrl, url);
+    }
+
+    String buildAppResponseAndCallExtraction(List<ListAppResponse> responses ){
+        var url = leonardoDao.extractWdsUrl(responses);
+        return url;
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java:82]: from IntelliJ, inline the variable:
```suggestion
      return leonardoDao.extractWdsUrl(responses);
```

File: service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    List<ListAppResponse> generateListAppResponse(String wdsKey, AppStatus wdsStatus, int count) {
+        List<ListAppResponse> responseList = new ArrayList<>();
+        var url = expectedUrl;
+        while(count != 0) {
+            ListAppResponse response = mock(ListAppResponse.class);
+            Map<String, String> proxyUrls = new HashMap<String, String>();
```
Comment (davidangb) [inline service/src/test/java/org/databiosphere/workspacedataservice/leonardo/LeonardoDaoTest.java:90]: from IntelliJ, let Java infer the HashMap types:
```suggestion
            Map<String, String> proxyUrls = new HashMap<>();
```

**Review by davidangb — COMMENTED** [review]
one comment in HttpLeonardoClientFactory that needs attention; I've also added suggestions from IntelliJ, but feel free to ignore them if you prefer the code as-is.

File: service/src/main/java/org/databiosphere/workspacedataservice/leonardo/HttpLeonardoClientFactory.java (hunk 1 above)
Comment (yuliadub) [reply service/src/main/java/org/databiosphere/workspacedataservice/leonardo/HttpLeonardoClientFactory.java:25]: wow, how was the code working if it was using the sam client??? thank you for spotting this!

**Review by davidangb — APPROVED** [review]

**Pull request merged**